src/mdns/mdns_avahi.cpp

```diff
@@ -533,79 +663,184 @@ otbrError PublisherAvahi::PublishService(const char *   aHostName,
 
-    return OTBR_ERROR_NONE;
+    otbrLog(OTBR_LOG_INFO, "[mdns] unpublish service %s.%s", aName, aType);
+    error = FreeGroup(serviceIt->mGroup);
+    mServices.erase(serviceIt);
+
+exit:
+    return error;
 }
 
 otbrError PublisherAvahi::PublishHost(const char *aName, const uint8_t *aAddress, uint8_t aAddressLength)
 {
-    OTBR_UNUSED_VARIABLE(aName);
-    OTBR_UNUSED_VARIABLE(aAddress);
-    OTBR_UNUSED_VARIABLE(aAddressLength);
+    otbrError       error      = OTBR_ERROR_NONE;
+    int             avahiError = 0;
+    Hosts::iterator hostIt     = mHosts.end();
+    std::string     fullHostName;
+    AvahiAddress    address;
 
-    VerifyOrDie(false, "PublishHost is not implemented with avahi");
+    VerifyOrExit(mState == State::kReady, errno = EAGAIN, error = OTBR_ERROR_ERRNO);
+    VerifyOrExit(mClient != nullptr, errno = EAGAIN, error = OTBR_ERROR_ERRNO);
+    VerifyOrExit(aName != nullptr, error = OTBR_ERROR_INVALID_ARGS);
+    VerifyOrExit(aAddress != nullptr, error = OTBR_ERROR_INVALID_ARGS);
+    VerifyOrExit(aAddressLength == sizeof(address.data.ipv6.address), error = OTBR_ERROR_INVALID_ARGS);
 
-    return OTBR_ERROR_NONE;
+    fullHostName = MakeFullName(aName);
+    hostIt       = FindHost(aName);
+
+    if (hostIt == mHosts.end())
+    {
+        SuccessOrExit(error = CreateHost(*mClient, aName, hostIt));
+    }
+    else if (memcmp(hostIt->mAddress.data.ipv6.address, aAddress, aAddressLength))
+    {
+        // Check if the update is necessary
+        VerifyOrExit(memcmp(hostIt->mAddress.data.ipv6.address, aAddress, aAddressLength));
+        SuccessOrExit(error = ResetGroup(hostIt->mGroup));
+    }
+    else
+    {
+        if (avahiError == 0 && mHostHandler != nullptr)
+        {
+            // The handler should be called even if the request can be processed synchronously
+            mHostHandler(aName, OTBR_ERROR_NONE, mHostHandlerContext);
+        }
+        ExitNow();
+    }
+
+    address.proto = AVAHI_PROTO_INET6;
+    memcpy(&address.data.ipv6.address[0], aAddress, aAddressLength);
+
+    otbrLog(OTBR_LOG_INFO, "[mdns] create host %s", aName);
+    avahiError = avahi_entry_group_add_address(hostIt->mGroup, AVAHI_IF_UNSPEC, AVAHI_PROTO_INET6,
+                                               AVAHI_PUBLISH_NO_REVERSE, fullHostName.c_str(), &address);
+    SuccessOrExit(avahiError);
+
+    otbrLog(OTBR_LOG_INFO, "[mdns] commit host %s", aName);
+    avahiError = avahi_entry_group_commit(hostIt->mGroup);
+    SuccessOrExit(avahiError);
+
+    hostIt->mAddress = address;
+
+exit:
+
+    if (avahiError)
+    {
+        error = OTBR_ERROR_MDNS;
+        otbrLog(OTBR_LOG_ERR, "Failed to publish host for avahi error: %s!", avahi_strerror(avahiError));
+    }
+    else if (error != OTBR_ERROR_NONE)
+    {
+        otbrLog(OTBR_LOG_ERR, "Failed to publish host: %s!",
+                error == OTBR_ERROR_ERRNO ? strerror(errno) : otbrErrorString(error));
+    }
+
+    if (error != OTBR_ERROR_NONE && hostIt != mHosts.end())
```

Comment:
Is there a case where `avahiError != 0` and `error != OTBR_ERROR_NONE`? If not, we can combine in above else-if statement?

Reply:
When `avahiError != 0`, `error` is assigned `OTBR_ERROR_MDNS` so after the `if-else-if- block it's possible.